pkg/devserver/server.go

```diff
@@ -31,25 +33,76 @@ func (s *Server) Start() error {
 	return nil
 }
 
+type PageTemplate struct {
+	Title   string
+	Content string
+}
+
+var pageTemplateContent = `<!DOCTYPE html>
+<html>
+	<head>
+		<title>{{ .Title }}</title>
+		<link rel="stylesheet" href="/assets/css/main.css">
+	</head>
+	<body>
+		<h1>{{ .Title }}</h1>
+		{{ .Content }}
+	</body>
+</html>`
+
+var pageTemplate *template.Template = nil
+
+type PageData struct {
+	Title   string
+	Content template.HTML
+}
+
+func init() {
+	var t, err = template.New("page").Parse(pageTemplateContent)
+	if err != nil {
+		panic(err)
+	}
+	pageTemplate = t
+}
+
+type Renderer struct{}
+
+func (r *Renderer) Render(w io.Writer, name string, data interface{}, c echo.Context) error {
+	if name != "page" {
+		return fmt.Errorf("unknown template: %s", name)
+	}
+
+	return pageTemplate.Execute(w, data)
+}
+
 func (s *Server) servePage(c echo.Context) error {
-	file, err := content.ParseFile(path.Join(s.config.ContentDir, c.Param("page")+".md"))
+	page := c.Param("page")
+
+	file, err := content.ParseFile(path.Join(s.config.ContentDir, page+".md"))
 	if err != nil {
 		return fmt.Errorf("error processing file: %w", err)
 	}
 
-	return c.HTMLBlob(http.StatusOK, file.ParsedContent)
+	return c.Render(http.StatusOK, "page", PageData{
+		Title:   page,
+		Content: template.HTML(string(file.ParsedContent)),
+	})
 }
 
 func NewServer(config Config) *Server {
 	echoInst := echo.New()
 
+	echoInst.Renderer = &Renderer{}
+
 	echoInst.Use(slogecho.New(slog.Default()))
 
 	server := &Server{
 		echoInst: echoInst,
 		config:   config,
 	}
 
+	echoInst.Static("/assets", "assets")
```

Comment:
Could the static assets be embedded into the binary, so it's a single self-contained binary?

To embed, you can use [embed](https://pkg.go.dev/embed), then for referencing them you can either:
- Replace this static function with the wrapped stdlib handler that works with an FS (example: https://github.com/nint8835/discord-whitelist/blob/main/pkg/server/server.go#L140)
- Use the static middleware, which exposes all static assets directly off the root (example: https://github.com/nint8835/discord-achievements/blob/main/pkg/app/app.go#L42)

Both take a http.FS, so for local development without needing to recompile you can swap the underlying FS between either an embedded one or a path read from the system (example: https://github.com/nint8835/discord-achievements/blob/main/pkg/app/app.go#L36-L40)

Reply:
should be done, wasn't so bad, the examples you gave of where you did this came in quite helpful!